### PR TITLE
Add ForOp (Functional for-loop operator)

### PR DESCRIPTION
**Pull request by segeljakt**

### CODE REVIEW

**Comment by segeljakt** [comment]
Currently the operator can only iterate over tensors, but we can extended it later to support range and zip iterators also. I think we can maybe wait with deriving the optimisations from the builtin loops.

File: arc-mlir/src/include/arc/Ops.td [hunk 1]
```diff
@@ -26,6 +26,7 @@
 #ifndef OP_BASE
 include "mlir/IR/OpBase.td"
 #endif // OP_BASE
+/* include "mlir/Transforms/LoopLikeInterface.td" */
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: This shouldn't be here

File: arc-mlir/src/include/arc/Ops.td [hunk 2]
```diff
@@ -258,6 +261,62 @@ def IfOp : Arc_Op<"if", [SingleBlockImplicitTerminator<"ArcBlockResultOp">]> {
   let hasCanonicalizer = 1;
 }
 
+def ArcYieldOp : Arc_Op<"yield", [Terminator, SameOperandsAndResultType]> {
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: This appears to almost the same as the `ArcBlockResultOp`, do we need both? `ArcYieldOp` is a better name so I prefer to keep the new operator. Looking at `ArcBlockResultOp` I see that it doesn't have any parent constraint, should you not require it to be a child of a loop? I think it should be possible to an 'or' constraint and use it both for if and loops.

File: arc-mlir/src/include/arc/Ops.td [hunk 3]
```diff
@@ -258,6 +261,62 @@ def IfOp : Arc_Op<"if", [SingleBlockImplicitTerminator<"ArcBlockResultOp">]> {
...
+    ```
+  }];
+
+  let arguments = (ins AnyType:$result);
+  let results = (outs AnyType);
+}
+
+def ForOp : Arc_Op<"for", [
+  Linear,
+/*   DeclareOpInterfaceMethods<LoopLikeOpInterface>, */
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: Forgotten comment?

File: arc-mlir/src/include/arc/Ops.td (hunk 2 above)
Comment (segeljakt) [reply arc-mlir/src/include/arc/Ops.td]: I'll try it out

File: arc-mlir/src/lib/Arc/Dialect.cpp [hunk 4]
```diff
@@ -152,6 +152,75 @@ LogicalResult IfOp::customVerify() {
   return mlir::success();
 }
 
+LogicalResult ArcYieldOp::customVerify() {
+  auto Op = this->getOperation();
+  auto ParentOp = Op->getParentOp();
+  auto Results = ParentOp->getResults();
+  auto Operands = Op->getOperands();
+
+  if (isa<IfOp>(ParentOp) || isa<ForOp>(ParentOp)) {
```
Comment (frej) [inline arc-mlir/src/lib/Arc/Dialect.cpp]: `isa<IfOp>(ParentOp) ||` shouldn't this go into the next patch if we are picky?

File: arc-mlir/src/lib/Arc/Dialect.cpp [hunk 5]
```diff
@@ -152,6 +152,75 @@ LogicalResult IfOp::customVerify() {
...
+      return Op->emitOpError() << "parent of yield must have same number of "
+                                  "results as the yield operands";
+    for (auto e : llvm::zip(Results, Operands)) {
+      if (std::get<0>(e).getType() != std::get<1>(e).getType())
+        return Op->emitOpError()
+               << "types mismatch between yield op and its parent";
+    }
+  } else {
+    return Op->emitOpError()
+           << "yield only terminates If, For or Parallel regions";
```
Comment (frej) [inline arc-mlir/src/lib/Arc/Dialect.cpp]: "If, " to the next patch?

**Comment by segeljakt** [comment]
This is ancient so I will close it

**Pull request closed without merging**